### PR TITLE
CI updates

### PR DESCRIPTION
**Pull request by honno**
* Checks the first two points of https://github.com/data-apis/array-api-tests/issues/235
  * Replace numpy.array_api with [array-api-strict](https://github.com/data-apis/array-api-strict)
  * Cover meta tests
* Removes `--ci` flag & splits the meta tests into a seperate folder `./meta_tests/`
  * Any test that wasn't in `./array_api_tests/meta/` was a CI test anywho
  * Default runs of the test suite should probably ignore the meta tests, as they're not very well structured and noisy for users
  * A seperate top-level folder avoids weird `conftest.py` hacks
* Also marks any test that doesn't use vectorization with an introduced `unvectorized` marker
  * Currently a no-op. My idea was to limit examples specifically on these on runtime, but I couldn't work that out easily so something for a follow-up PR.
  * Generally good for us to denote which tests could possibly be improved on performance-wise.

### CODE REVIEW

File: array_api_tests/conftest.py [hunk 1]
```diff
@@ -54,11 +54,6 @@ def pytest_addoption(parser):
         help="disable testing functions with output shapes dependent on input",
     )
     # CI
-    parser.addoption(
```
Comment (asmeurer) [inline array_api_tests/conftest.py:57]: Everyone using the test suite on CI is using this option. Should maybe leave this in for backwards compatibility.

File: .github/workflows/numpy.yml [hunk 2]
```diff
@@ -22,10 +22,12 @@ jobs:
...
-        python -m pip install numpy==1.26.2
+        python -m pip install array-api-strict
         python -m pip install -r requirements.txt
     - name: Run the test suite
       env:
-        ARRAY_API_TESTS_MODULE: numpy.array_api
+        ARRAY_API_TESTS_MODULE: array_api_strict
       run: |
-        pytest -v -rxXfE --ci --skips-file numpy-skips.txt
+        pytest -v -rxXfE --skips-file numpy-skips.txt array_api_tests/
```
Comment (asmeurer) [inline .github/workflows/numpy.yml:31]: This is the skips file I use in array-api-strict. https://github.com/data-apis/array-api-strict/blob/main/array-api-tests-xfails.txt. There's one special case test that fails in NumPy and one test that's a bug in the test suite.

**Comment by asmeurer** [comment]
I get this error when I try to use this locally

```
$ARRAY_API_TESTS_MODULE=numpy.array_api pytest
=========================================================================== test session starts ===========================================================================
platform darwin -- Python 3.11.7, pytest-8.0.0, pluggy-1.4.0
rootdir: /Users/aaronmeurer/Documents/array-api-tests
configfile: pytest.ini
plugins: pudb-0.7.0, hypothesis-6.99.2, json-report-1.5.0, cov-4.1.0, doctestplus-1.1.0, typeguard-4.1.5, flakes-4.0.5
collected 69 items / 1 error
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1645, in getoption
INTERNALERROR>     val = getattr(self.option, name)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR> AttributeError: 'Namespace' object has no attribute 'skips_file'
INTERNALERROR>
INTERNALERROR> The above exception was the direct cause of the following exception:
INTERNALERROR>
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/main.py", line 272, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>                          ^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/main.py", line 325, in _main
INTERNALERROR>     config.hook.pytest_collection(session=session)
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_hooks.py", line 501, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_manager.py", line 119, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 138, in _multicall
INTERNALERROR>     raise exception.with_traceback(exception.__traceback__)
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/logging.py", line 783, in pytest_collection
INTERNALERROR>     return (yield)
INTERNALERROR>             ^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/warnings.py", line 118, in pytest_collection
INTERNALERROR>     return (yield)
INTERNALERROR>             ^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1365, in pytest_collection
INTERNALERROR>     return (yield)
INTERNALERROR>             ^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 102, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/main.py", line 336, in pytest_collection
INTERNALERROR>     session.perform_collect()
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/main.py", line 802, in perform_collect
INTERNALERROR>     hook.pytest_collection_modifyitems(
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_hooks.py", line 501, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_manager.py", line 119, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 138, in _multicall
INTERNALERROR>     raise exception.with_traceback(exception.__traceback__)
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/cacheprovider.py", line 413, in pytest_collection_modifyitems
INTERNALERROR>     res = yield
INTERNALERROR>           ^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/cacheprovider.py", line 339, in pytest_collection_modifyitems
INTERNALERROR>     res = yield
INTERNALERROR>           ^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/pluggy/_callers.py", line 102, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/Documents/array-api-tests/array_api_tests/conftest.py", line 114, in pytest_collection_modifyitems
INTERNALERROR>     skips_file = skips_path = config.getoption('--skips-file')
INTERNALERROR>                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/Users/aaronmeurer/miniconda3/envs/array-apis/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1656, in getoption
INTERNALERROR>     raise ValueError(f"no option named {name!r}") from e
INTERNALERROR> ValueError: no option named 'skips_file'
```

File: .github/workflows/numpy.yml
Comment (asmeurer) [inline .github/workflows/numpy.yml:1]: Should change the name of the file and job name too.

File: array_api_tests/conftest.py (hunk 1 above)
Comment (honno) [reply array_api_tests/conftest.py:57]: Yeah that makes sense, although a persisting concern of mine is the test suite is inherently pretty complex so it's worth being a bit ruthless in simplifying things wherever possible :thinking:

**Comment by honno** [comment]
> I get this error when I try to use this locally...


Ah I think this is why `conftest.py` exists in the top-level then :sweat_smile:

File: array_api_tests/conftest.py (hunk 1 above)
Comment (asmeurer) [reply array_api_tests/conftest.py]: We can deprecate it, although most people won't see the warning. I'd rather not break existing users if we don't have to.

**Comment by asmeurer** [comment]
Tests seem to work locally now.

**Comment by asmeurer** [comment]
> Ah I think this is why conftest.py exists in the top-level then 😅 Reverted that change.

pytest has some seriously annoying limitations like this. It's a good thing we aren't trying to distribute the test suite as a package because this sort of thing would make doing that extremely difficult.

File: array_api_tests/conftest.py (hunk 1 above)
Comment (honno) [reply array_api_tests/conftest.py]: Deprecated instead, does feel right now!

**Comment by honno** [comment]
One more thing, I've figured out how to limit examples for the tests marked with `unvectorized` in the `pytest_collection_modifyitems` hook, so have added that functionality to `conftest.py`. There's room for improvement and I'll explore a way to do it without using private (if stable) Hypothesis APIs, but should be a nice quick win, and I should of guarded nicely if Hypothesis in the future complains at what we're doing.

**Comment by asmeurer** [comment]
Feel free to merge this whenever you feel it is ready.

**Comment by asmeurer** [comment]
This broke the array-api-compat CI. Fix at https://github.com/data-apis/array-api-tests/pull/245

**Comment by asmeurer** [comment]
How do you feel about adding these as required checks in the branch protection rules? Admins would still be able to bypass the checks if necessary, but it would making merging a PR that breaks the tests much less likely.

**Comment by honno** [comment]
> How do you feel about adding these as required checks in the branch protection rules? Admins would still be able to bypass the checks if necessary, but it would making merging a PR that breaks the tests much less likely.

Mind my ignorance on branch protection rules—what would we be adding as required checks exactly? The tests pasing? Or new tests for `array-api-compat`?

**Comment by asmeurer** [comment]
The tests for this repo. It isn't possible to require checks across repos. 

I went ahead and enabled it. You can go to https://github.com/data-apis/array-api-tests/settings/branch_protection_rules/22738169 to change the settings.

**Pull request merged**